File: packages/db/src/queries/review.ts
import { and, eq, inArray, isNull, or, type SQL } from "drizzle-orm";

import { transactions } from "../schema/ledger";

/**
 * O que faz um lançamento precisar de olho humano — a regra, em um lugar só.
 *
 * Ela vivia dentro de `apps/web/lib/review.ts`, atrás de `server-only`, o que
 * teve duas consequências. A primeira era conhecida: o badge da navegação e a
 * tela da fila precisam concordar sempre, senão o badge diz "3" e a tela abre
 * vazia. A segunda não era: **o agente não alcançava a fila**. A pessoa
 * perguntava "o que está para revisar sem categoria" e a Clara não tinha porta
 * nenhuma para essa informação — a resposta existia, na tela ao lado, e ela
 * respondia que não sabia.
 *
 * Copiar o predicado para o pacote do agente resolveria o acesso e criaria o
 * problema pior: duas definições de "precisa de revisão" divergindo com o
 * tempo. Então ele mora aqui, onde os dois chegam.
 */
export type ReviewReason = "sem_categoria" | "confianca_baixa" | "sem_comerciante";

/**
 * Gasto que precisa de categoria — pagamento e ajuste ficam fora.
 *
 * Pagamento de fatura não é gasto a classificar, e ajuste de saldo tampouco:
 * contá-los inflava o indicador com trabalho que não existe (nas duas primeiras
 * faturas reais, 4 dos 9 "sem categoria" eram um ou outro).
 *
 * Mora aqui, e não em cada consumidor, porque quatro lugares derivavam este
 * recorte por conta própria — o estado do razão, a triagem do guarda-livros, a
 * simulação das regras aprendidas e a fila de revisão — e dois deles haviam
 * divergido, contando pagamento. Dois números para a mesma pergunta, no mesmo
 * contexto, é o começo de uma resposta que pede desculpa.
 */
export const CATEGORIZABLE_KINDS = ["purchase", "refund", "fee"] as const;

function isCategorizable(kind: string): boolean {
  return (CATEGORIZABLE_KINDS as readonly string[]).includes(kind);
}

export function uncategorizedSpendCondition(): SQL | undefined {
  return and(isNull(transactions.category), inArray(transactions.kind, [...CATEGORIZABLE_KINDS]));
}

/**
 * Três motivos, e cada um dói de um jeito:
 *
 *  - **sem categoria** — toda análise por categoria fica com um buraco;
 *  - **confiança baixa** — o extrator avisou que pode ter lido errado, e o
 *    valor conta como gasto de qualquer forma;
 *  - **sem comerciante** — a linha não se agrupa com nada, nem em recorrência
 *    nem em regra aprendida.
 *
 * Sai da fila quem foi ATESTADO (`reviewedAt`), não quem "parece resolvido": a
 * conclusão mais comum de uma revisão é "a leitura já estava certa", e sem o
 * atestado esses itens voltariam para sempre.
 */
export function needsReviewCondition(): SQL | undefined {
  return and(isNull(transactions.reviewedAt), reviewFlagsCondition());
}

/**
 * Os mesmos três motivos, SEM o filtro do atestado.
 *
 * A separação existe porque o atestado e o motivo respondem a perguntas
 * diferentes, e confundi-las produziu uma resposta impossível de entender:
 * *"existem 2 itens sem categoria"* — dito pelo estado do razão, que não olha
 * `reviewedAt` — seguido de *"a consulta não retornou os itens"*, porque a fila
 * de revisão só enxerga o que ninguém atestou ainda. Um lançamento atestado com
 * a conclusão "a leitura estava certa, mas nenhuma categoria descreve isto"
 * está exatamente nesse limbo: conta no indicador e não aparece na fila.
 *
 * Com o predicado separado, a fila continua sendo a fila (o que espera olho
 * humano) e ainda assim alcança o item atestado quando a pergunta é sobre ele.
 */
export function reviewFlagsCondition(): SQL | undefined {
  return or(
    // "Sem categoria" aqui é o MESMO recorte do indicador do turno: pagamento
    // de fatura não espera categoria nenhuma, e marcá-lo assim tinha duas
    // consequências, uma pior que a outra. A pequena: ele nunca saía da fila,
    // porque a única coisa que o tiraria — uma categoria — jamais deveria
    // chegar. A grande: a fila devolvia 3 itens "sem categoria" onde o estado
    // do razão contava 2, e a resposta apresentava uma lista que não fechava
    // com o número dito na frase anterior.
    uncategorizedSpendCondition(),
    eq(transactions.extractionConfidence, "baixa"),
    isNull(transactions.merchant),
  );
}

/**
 * Por que ESTE lançamento está na fila. Pode ser mais de um motivo.
 *
 * Espelha `reviewFlagsCondition()` linha por linha — inclusive o `kind`. Quando
 * as duas discordavam, a consulta trazia a linha e a lista de motivos vinha
 * vazia: um item na fila sem nenhuma razão para estar nela.
 */
export function reviewReasonsFor(transaction: {
  category: string | null;
  merchant: string | null;
  kind: string;
  extractionConfidence: "alta" | "media" | "baixa";
}): ReviewReason[] {
  const reasons: ReviewReason[] = [];
  if (transaction.category === null && isCategorizable(transaction.kind)) {
    reasons.push("sem_categoria");
  }
  if (transaction.extractionConfidence === "baixa") reasons.push("confianca_baixa");
  if (transaction.merchant === null) reasons.push("sem_comerciante");
  return reasons;
}

File: services/agent/clara/db/queries/review.py
"""O que faz um lançamento precisar de olho humano — o predicado, em um lugar só.

Porta de `packages/db/src/queries/review.ts`. Consumido pela fila de revisão,
pelo snapshot do razão (FR-021) e pela triagem do categorizador — os três
precisam concordar sempre, ou a conversa afirma "há N itens sem categoria" e
não consegue listá-los (FR-020).
"""

from __future__ import annotations

from typing import Literal

from sqlalchemy import ColumnElement, and_, or_

from clara.db.models import Transaction

ReviewReason = Literal["sem_categoria", "confianca_baixa", "sem_comerciante"]

# Pagamento de fatura e ajuste de saldo não esperam categoria nenhuma —
# contá-los infla o indicador com trabalho que não existe.
CATEGORIZABLE_KINDS = ("purchase", "refund", "fee")


def _is_categorizable(kind: str) -> bool:
    return kind in CATEGORIZABLE_KINDS


def uncategorized_spend_condition() -> ColumnElement[bool]:
    """Gasto que precisa de categoria — pagamento e ajuste ficam fora."""
    return and_(
        Transaction.category.is_(None),
        Transaction.kind.in_(CATEGORIZABLE_KINDS),
    )


def review_flags_condition() -> ColumnElement[bool]:
    """Os três motivos, SEM o filtro do atestado — separado de propósito.

    Confundir atestado com motivo produz uma resposta impossível de entender:
    "existem 2 sem categoria" (o snapshot, que não olha `reviewed_at`) seguido
    de "a fila não retornou nada" (que só enxerga o que ninguém atestou ainda).
    """
    return or_(
        uncategorized_spend_condition(),
        Transaction.extraction_confidence == "baixa",
        Transaction.merchant.is_(None),
    )


def needs_review_condition() -> ColumnElement[bool]:
    """Sai da fila quem foi ATESTADO (`reviewed_at`), não quem "parece resolvido"."""
    return and_(Transaction.reviewed_at.is_(None), review_flags_condition())


def review_reasons_for(
    *,
    category: str | None,
    merchant: str | None,
    kind: str,
    extraction_confidence: Literal["alta", "media", "baixa"],
) -> list[ReviewReason]:
    """Por que ESTE lançamento está na fila — pode ser mais de um motivo.

    Espelha `review_flags_condition()` linha por linha, inclusive o `kind`,
    para que a linha nunca apareça na fila sem nenhum motivo listado.
    """
    reasons: list[ReviewReason] = []
    if category is None and _is_categorizable(kind):
        reasons.append("sem_categoria")
    if extraction_confidence == "baixa":
        reasons.append("confianca_baixa")
    if merchant is None:
        reasons.append("sem_comerciante")
    return reasons
